apps/experiments/models.py

```diff
@@ -422,6 +423,20 @@ def get_api_url(self):
     def get_absolute_url(self):
         return reverse("experiments:single_experiment_home", args=[self.team.slug, self.id])
 
+    @property
+    def only_whitelisted_participants_allowed(self):
+        """
+        Whether or not a bot should be limited to whitelisted participants is inferred by the presence of a whitelist
+        """
+        return len(self.participant_whitelist) > 0
+
+    def get_whitelisted_participant_identifiers(self):
+        whitelist = []
+        team_user_emails = self.team.members.all().values_list("email", flat=True)
+        whitelist.extend(list(team_user_emails))
+        whitelist.extend(self.participant_whitelist)
+        return whitelist
```

Comment:
Rather than doing a query every time you could first check the whitelist and then only query the DB if necessary. You also don't need to load all team members, just do an 'exists' query with filtering.

Reply:
Good suggestion, I'll update

Reply:
https://github.com/dimagi/open-chat-studio/pull/625/commits/af8473ae0cea605473ed4794f1d5dc8848ed9754